src/meta/src/barrier/progress.rs

```diff
@@ -436,71 +420,70 @@ impl CreateMviewProgressTracker {
+    pub(super) fn take_finished_jobs(&mut self) -> Vec<TrackingJob> {
         tracing::trace!(finished_jobs=?self.finished_jobs, progress_map=?self.progress_map, "finishing jobs");
-        for job in self
-            .finished_jobs
-            .extract_if(|job| checkpoint || !job.is_checkpoint_required())
-        {
-            // The command is ready to finish. We can now call `pre_finish`.
-            job.pre_finish().await?;
-        }
-        Ok(!self.finished_jobs.is_empty())
+        take(&mut self.finished_jobs)
+    }
+
+    pub(super) fn has_pending_finished_jobs(&self) -> bool {
+        !self.finished_jobs.is_empty()
     }
 
     pub(super) fn cancel_command(&mut self, id: TableId) {
         let _ = self.progress_map.remove(&id);
-        self.finished_jobs
-            .retain(|x| x.table_to_create() != Some(id));
+        self.finished_jobs.retain(|x| x.table_to_create() != id);
         self.actor_map.retain(|_, table_id| *table_id != id);
     }
 
     /// Notify all tracked commands that error encountered and clear them.
-    pub async fn abort_all(&mut self, err: &MetaError, context: &GlobalBarrierManagerContext) {
+    pub fn abort_all(&mut self) {
         self.actor_map.clear();
         self.finished_jobs.clear();
         self.progress_map.clear();
-        match &context.metadata_manager {
-            MetadataManager::V1(mgr) => {
-                mgr.notify_finish_failed(err).await;
-            }
-            MetadataManager::V2(mgr) => {
-                mgr.notify_finish_failed(err).await;
-            }
-        }
     }
 
     /// Add a new create-mview DDL command to track.
     ///
     /// If the actors to track is empty, return the given command as it can be finished immediately.
     pub fn add(
         &mut self,
-        command: TrackingCommand,
+        command_ctx: &Arc<CommandContext>,
         version_stats: &HummockVersionStats,
     ) -> Option<TrackingJob> {
-        let actors = command.context.actors_to_track();
-        if actors.is_empty() {
-            // The command can be finished immediately.
-            return Some(TrackingJob::New(command));
-        }
+        let (info, actors, replace_table_info) = if let Command::CreateStreamingJob {
+            info,
+            replace_table,
+        } = &command_ctx.command
+        {
+            let CreateStreamingJobCommandInfo {
+                table_fragments, ..
+            } = info;
+            let actors = table_fragments.tracking_progress_actor_ids();
+            if actors.is_empty() {
+                // The command can be finished immediately.
+                return Some(TrackingJob::New(TrackingCommand {
+                    info: info.clone(),
+                    replace_table_info: replace_table.clone(),
+                }));
+            }
+            (info.clone(), actors, replace_table.clone())
+        } else {
+            return None;
```

Comment:
Will this be a problem? Previously if not a `Command::CreateStreamingJob`, we will still return `Some(...)`.

Because the caller of `add` will only push the command to the `finished` queue when the result from `add` is `Some(...)`.

Reply:
Here: https://github.com/risingwavelabs/risingwave/pull/17742/files#r1684104948

Reply:
For command returned with `Some(_)`, they will only be called `pre_finish`. However, for command not `CreateStreamingJob`, `pre_finish` simply returns `Ok(())`.